crates/api/src/lib.rs

```diff
@@ -150,35 +157,77 @@ pub(crate) fn build_totp_2fa(
   .with_lemmy_type(LemmyErrorType::CouldntGenerateTotp)
 }
 
-/// Since removals and bans are only federated for local users,
-/// you also need to send bans for their content to local communities.
+/// Site bans are only federated for local users.
+/// This is a problem, because site-banning non-local users will still leave content
+/// they've posted to our local communities, on other servers.
+///
+/// So when doing a site ban for a non-local user, you need to federate/send a
+/// community ban for every local community they've participated in.
 /// See https://github.com/LemmyNet/lemmy/issues/4118
 #[tracing::instrument(skip_all)]
-pub(crate) async fn send_bans_and_removals_to_local_communities(
+pub(crate) async fn ban_nonlocal_user_from_local_communities(
   local_user_view: &LocalUserView,
   target: &Person,
+  ban: bool,
   reason: &Option<String>,
   remove_data: &Option<bool>,
+  expires: &Option<i64>,
```

Comment:
If its a temporary site ban then the user would get banned from local communities and never unbanned, right?

Reply:
Also instead of running this logic on the instance where the ban originates, we could move it to the code which [receives remote site bans](https://github.com/LemmyNet/lemmy/blob/main/crates/apub/src/activities/block/block_user.rs#L157). Then it would only have to ban the user from all known communities of that instance. This would reduce the amount of activities that have to get federated. However it wont work if a community gets federated later. Though with this implementation, the ban also wont federate to instances where the local communities dont have any followers.

Either way the core problem wont be solved. We only federate bans via activities, so if that doesnt get delivered or an instance fetches the community later, it wont know about the ban. A proper solution would be similar to https://github.com/LemmyNet/lemmy/pull/4475, federating a collection of users who are banned from each instance/community so that it can also be synchronized later. And for site bans we need to store them in a separate table as `table site_ban (instance_id, person_id)`, to properly track which instance banned the user. All rather complex, so we can use this PR as a temporary, imperfect solution.

Reply:
> If its a temporary site ban then the user would get banned from local communities and never unbanned, right?

If you look below, the community ban form also includes that expires. So the site ban, and the community bans, would all expire at the same time.

> Also if ban = false, any existing community bans for the user would be removed. Seems safer to remove these two params and use it only in case of permanent site ban.

That's true, but I think its still a good idea to federate community unbans. If we didn't run this code for unbans, then if someone was permabanned, then unbanned, their community bans would still be there. 

Its a limitation, since community mods *might* have wanted that user banned, but I think overall its still better than leaving all the bans in place.

> Also instead of running this logic on the instance where the ban originates, we could move it to the code which [receives remote site bans](https://github.com/LemmyNet/lemmy/blob/main/crates/apub/src/activities/block/block_user.rs#L157). Then it would only have to ban the user from all known communities of that instance. This would reduce the amount of activities that have to get federated. However it wont work if a community gets federated later. Though with this implementation, the ban also wont federate to instances where the local communities dont have any followers.

That's up to you, but this code currently does work, and the API tests are passing. I don't think it makes much sense to federate out local site bans.